### ISSUE TITLE
Size matters

### ISSUE THREAD

**Issue by som-snytt** [scalapuzzlers/scalapuzzlers.github.com#146]
I don't know if this has come up.

https://issues.scala-lang.org/browse/SI-9738

Which combos work?

```
Set(2, 1).reduce(math.max)

Set(2L, 1).reduce(math.max)

Set(2L, 1L).reduce(math.max)
```

Those aren't funny numbers, so that would have to be improved, of course.

"42 and 17 had just finished working out at the gym..."

**Comment by demobox** [comment]
"Size matters"

> Ha! We were considering that title for http://scalapuzzlers.com/#pzzlr-046, but decided to go with something a bit more neutral ;-)

This one certainly looks funky, thanks! The only one I can think of that may get close to this is http://scalapuzzlers.com/#pzzlr-023 (which in the book was changed quite a bit to be one about the interplay between implicit conversions and shape-based overloading resolution); as far as I can make out so far, the underlying mechanism here is different.

From the issue comments I'm not quite clear yet whether this is "as specified", or actually a compiler bug?

**Comment by som-snytt** [comment]
Definitely as specified, or as loosely specified. It might be a case where inference is underspecified, since the expected type of the eta-expanded func could have disqualified some overloads.

That's why I suggest that a debug mode -Xlint could explain that eta-expansion happens before the type param is solved for. That's why it picks AnyVal for the type arg.

On the one hand, I don't envy your puzzler job; but at least you get to laugh all the way to bank.

**Comment by demobox** [comment]
> -Xlint could explain that eta-expansion happens before the type param is solved for. That's why it picks AnyVal for the type arg.

Ah, thanks for explaining! A definite puzzler in that case, I think. Could be used in combination with
```
scala> Seq(1L, 2L).reduce(math.max(_, _))
res3: Long = 2
```
but then I suspect it might look too much like http://scalapuzzlers.com/#pzzlr-062 or http://scalapuzzlers.com/#pzzlr-022, which are also based on (other) differences between partial application and eta-expansion.

From your explanation, I would imagine this would work for single-parameter cases, too - will have a look around to see if there's anything there.

**Comment by demobox** [comment]
> Overloading is picking the most specific first, which is possible because of widening conversions. 

I'm assuming that comment (from [the ticket](https://issues.scala-lang.org/browse/SI-9738?focusedCommentId=74387&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-74387)) is explaining why it happens to work for `Int`? I guess that's because, even though `Int` is not at the bottom of the "weakly confroms" numeric widening hierarchy, it's the lowest one for which `math.max` is defined?

> eta-expansion happens before the type param is solved for. That's why it picks AnyVal for the type arg.

Just going back to that briefly: why `AnyVal`, seeing as that's not the type of any of the overloaded `math.max` methods? Triggering eta-expansion without the context (if I'm doing that correctly) chooses `Int` rather than `AnyVal`, as far as I can see?

```
scala> math.max _
res35: (Int, Int) => Int = <function2>
```

If I'm interpreting the error message correctly, it's looking for an `(AnyVal, AnyVal) => AnyVal` function because it has determined `AnyVal` to be the solution of type equation for `reduce`'s type parameter?

I think I'll need to dig in more detail into the definition of those constraint systems to try to understand _why_, though. Intuitively (not a particularly realistic starting point in this area of the spec, perhaps) if it hasn't yet determined the type of `math.max`, I would probably have guessed that it would first try the type of the set's elements?

If the type of `math.max` has already been determined as `(Int, Int) => Int` (following the "lowest available in the weakly conforming hierarchy" reasoning), I would more likely have expected an error along the lines of "can't reduce Longs with a function that takes Ints".

Is that perhaps the actual error that the compiler is trying to report, and the message is odd? Or if the error is indeed accurate, is it because this part of the resolution (i.e. the determination of the type parameter) doesn't use weak conformance, and is instead only looking at the type hierarchy relationship?

Scanning 6.26.6. ("Local Type Inference, Case 3: Methods"), I'm guessing that might be the case, since while weak conformance is mentioned in connection with _argument_ types, it does not seem to apply to the _result_ type?

**Comment by som-snytt** [comment]
I haven't looked at typer-debug output yet. We'd like it to pick the type arg at the same time as it picks max. Instead, it picks max(Int,Int) as more specific than max(Long,Long). If it's a set of Long, then the type arg to reduce is AnyVal, its best shot, only because AnyVal >:> Long.

**Comment by demobox** [comment]
> If it's a set of Long, then the type arg to reduce is AnyVal, its best shot, only because AnyVal >:> Long.

So at the point it picks `AnyVal`, it's already picked the `(Int, Int)` variant? If so, why then pick `AnyVal` - seeing as the result won't compile? Or is it **not** actually choosing AnyVal, but instead reporting an error that's basically "Unable to determine the type parameter for argument (Int, Int) => Int"?

Thanks for the further comments. Curious to hear what the typer-debug output produces!

**Comment by som-snytt** [comment]
```
$ scalam -Ytyper-debug -Ylog:typer
Welcome to Scala 2.12.0-RC1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_92).
Type in expressions for evaluation. Or try :help.

scala> Set(42L, 17).reduce(math.max)
|-- $line3 EXPRmode-POLYmode-QUALmode (site: package <root>) 
|    \-> $line3.type
|-- object $read BYVALmode-EXPRmode (site: package $line3) 
|    |-- super EXPRmode-POLYmode-QUALmode (silent: <init> in $read) 
|    |    |-- this EXPRmode (silent: <init> in $read) 
|    |    |    \-> $line3.$read.type
|    |    \-> $line3.$read.type
|    |-- object $iw BYVALmode-EXPRmode (site: object $read) 
|    |    |-- super EXPRmode-POLYmode-QUALmode (silent: <init> in $iw) 
|    |    |    |-- this EXPRmode (silent: <init> in $iw) 
|    |    |    |    \-> $line3.$read.$iw.type
|    |    |    \-> AnyRef
|    |    |-- object $iw BYVALmode-EXPRmode (site: object $iw) 
|    |    |    |-- super EXPRmode-POLYmode-QUALmode (silent: <init> in $iw) 
|    |    |    |    |-- this EXPRmode (silent: <init> in $iw) 
|    |    |    |    |    \-> $line3.$read.$iw.$iw.type
|    |    |    |    \-> AnyRef
|    |    |    |-- Set(42L, 17).reduce(math.max) EXPRmode (site: value res0  in $iw) 
|    |    |    |    |-- Set(42L, 17).reduce BYVALmode-EXPRmode-FUNmode-POLYmode (silent: value res0  in $iw) 
|    |    |    |    |    |-- Set(42L, 17) EXPRmode-POLYmode-QUALmode (silent: value res0  in $iw) 
|    |    |    |    |    |    |-- Set BYVALmode-EXPRmode-FUNmode-POLYmode (silent: value res0  in $iw) 
|    |    |    |    |    |    |    |-- scala.Predef.Set.apply BYVALmode-EXPRmode-FUNmode-POLYmode (silent: value res0  in $iw) 
|    |    |    |    |    |    |    |    [adapt] [A](elems: A*)CC[A] adapted to [A](elems: A*)CC[A]
|    |    |    |    |    |    |    |    \-> (elems: A*)scala.collection.immutable.Set[A]
|    |    |    |    |    |    |    [adapt] => scala.collection.immutable.Set.type adapted to [A](elems: A*)CC[A]
|    |    |    |    |    |    |    \-> (elems: A*)scala.collection.immutable.Set[A]
|    |    |    |    |    |    |-- 42L BYVALmode-EXPRmode-POLYmode (silent: value res0  in $iw) 
|    |    |    |    |    |    |    \-> Long(42L)
|    |    |    |    |    |    |-- 17 BYVALmode-EXPRmode-POLYmode (silent: value res0  in $iw) 
|    |    |    |    |    |    |    \-> Int(17)
|    |    |    |    |    |    solving for (A: ?A)
|    |    |    |    |    |    \-> scala.collection.immutable.Set[Long]
|    |    |    |    |    [adapt] [A1 >: A](op: (A1, A1) => A1)A1 adapted to [A1 >: A](op: (A1, A1) => A1)A1
|    |    |    |    |    \-> (op: (A1, A1) => A1)A1
|    |    |    |    |-- math.max : pt=(?, ?) => ? BYVALmode-EXPRmode-POLYmode (site: value res0  in $iw) 
|    |    |    |    |    |-- math EXPRmode-POLYmode-QUALmode (site: value res0  in $iw) 
|    |    |    |    |    |    \-> math.type
|    |    |    |    |    |-- scala.math.`package` EXPRmode-POLYmode-QUALmode (silent: value res0  in $iw) 
|    |    |    |    |    |    \-> math.type
|    |    |    |    |    |-- { ((x: Int, y: Int) => scala.math.`package`.max(x, y)) } : pt=(?, ?) => ? BYVALmode-EXPRmode-POLYmode (site: value res0  in $iw) 
|    |    |    |    |    |    |-- ((x: Int, y: Int) => scala.math.`package`.max(x, y)) : pt=(?, ?) => ? BYVALmode-EXPRmode-POLYmode (site: value res0  in $iw) 
|    |    |    |    |    |    |    |-- scala.math.`package`.max(x, y) EXPRmode (site: value $anonfun in $iw) 
|    |    |    |    |    |    |    |    |-- scala.math.`package`.max BYVALmode-EXPRmode-FUNmode-POLYmode (silent: value $anonfun in $iw) 
|    |    |    |    |    |    |    |    |    \-> (x: Double, y: Double)Double <and> (x: Float, y: Float)Float <and> (x: Long, y: Long)Long <and> (x: Int, y: Int)Int
|    |    |    |    |    |    |    |    |-- x BYVALmode-EXPRmode (silent: value $anonfun in $iw) 
|    |    |    |    |    |    |    |    |    \-> Int
|    |    |    |    |    |    |    |    |-- y BYVALmode-EXPRmode (silent: value $anonfun in $iw) 
|    |    |    |    |    |    |    |    |    \-> Int
|    |    |    |    |    |    |    |    \-> Int
|    |    |    |    |    |    |    \-> (Int, Int) => Int
|    |    |    |    |    |    \-> (Int, Int) => Int
|    |    |    |    |    [adapt] (x: Int, y: Int)Int adapted to { ((x: Int, y: Int) => scala.math.`package`.max(x, y)) } based on pt (?, ?) => ?
|    |    |    |    |    \-> (Int, Int) => Int
|    |    |    |    solving for (A1: ?A1)
[log typer] Inferred type for A1 in TraversableOnce[+A]#reduce[A1 >: A] (AnyVal) does not conform to bounds:  >: (Long, Int) <: Int _= AnyVal
|    |    |    |    [search #1] start `{
|    |    |    |      ((x: Int, y: Int) => scala.math.`package`.max(x, y))
|    |    |    |    }`, searching for adaptation to pt=((Int, Int) => Int) => ((AnyVal, AnyVal) => AnyVal) (silent: value res0  in $iw) implicits disabled
|    |    |    |    [search #2] start `{
|    |    |    |      ((x: Int, y: Int) => scala.math.`package`.max(x, y))
|    |    |    |    }`, searching for adaptation to pt=(=> (Int, Int) => Int) => ((AnyVal, AnyVal) => AnyVal) (silent: value res0  in $iw) implicits disabled
<console>:12: error: type mismatch;
 found   : (Int, Int) => Int
 required: (AnyVal, AnyVal) => AnyVal
       Set(42L, 17).reduce(math.max)
                                ^
|    |    |    |    \-> AnyVal
|    |    |    |-- AnyVal BYVALmode-EXPRmode (site: object $iw) 
|    |    |    |    \-> [val res0] AnyVal
|    |    |    |-- def res0 BYVALmode-EXPRmode (site: object $iw) 
|    |    |    |    |-- AnyVal : pt=AnyVal EXPRmode (site: value res0 in $iw) 
|    |    |    |    |    |-- type EXPRmode-POLYmode-QUALmode (site: value res0 in $iw) 
|    |    |    |    |    |    \-> $line3.$read.$iw.$iw.type
|    |    |    |    |    \-> AnyVal
|    |    |    |    \-> [def res0] => AnyVal
|    |    |    \-> [object $iw] type
|    |    \-> [object $iw] type
|    \-> [object $read] $line3.$read.type


```

**Comment by demobox** [comment]
@som-snytt Haven't forgotten this, just rushing around organizing an event at the moment. Will try to get back to this in a couple of weeks, once things are a bit quieter again...

**Comment by som-snytt** [comment]
I forgot it.

**Comment by demobox** [comment]
@som-snytt Taking a short break from grappling with politics and trying to wrap my head around this again. I think I have a handle on most of it, but I'm still missing a few pieces. I guess it boils down to:

1. Is the compiler using the expected type of `(T, T) => T` for some `T >: Long` (assuming that is the correct type) during overloading resolution for `math.max` and, if so, how is it concluding that `(Int, Int) => Int` is compatible with that type?
1. If the compiler does not consider an expected type to be available when determining which `math.max` variant to use, why not?

The missing pieces on the way to that question are:

* Is eta expansion kicking in _before_ or _after_ overload resolution has picked the specific version of `math.max` to use? From the spec, it would seem that eta expansion could only proceed once you know the specific expression of method type that is given, but from Paul's comment in [the ticket](https://issues.scala-lang.org/browse/SI-9738) and the debug output above, it looks like eta expansion may be happening before the compiler has picked the overloaded variant of `math.max`?

* Assuming the eta expansion has not yet happened, but that the compiler is applying overloading resolution to some as-yet-undetermined expression of method type, I would guess that the third case [in the spec](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#overloading-resolution) should apply:

> Assume finally that e does not appear as a function in either an application or a type application. If an expected type is given, let B be the set of those alternatives in A which are compatible to it. Otherwise, let B be the same as A. In this last case we choose the most specific alternative among all alternatives in B. It is an error if there is no alternative in B which is more specific than all other alternatives in B.

If that is indeed the overloading resolution variant that is applied, does it make use of the expected type of `(T, T) => T` for some `T >: Long`? Or is there not considered to be an expected type here? If the expected type _is_ used, is `(Int, Int) => Int` considered to be compatible with `(T, T) => T` for some `T >: Long`?

**Comment by demobox** [comment]
> 2. If the compiler does not consider an expected type to be available when determining which `math.max` variant to use, why not?

@som-snytt: This question is also related to your [code sample in the ticket](https://issues.scala-lang.org/browse/SI-9738?focusedCommentId=74387&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-74387):

```
scala> object X { def f(i: Int, j: Int) = i + j ; def f(s: String, t: String) = s + t }
scala> Set(1,2) reduce X.f
<console>:13: error: ambiguous reference to overloaded definition,
both method f in object X of type (s: String, t: String)String
and  method f in object X of type (i: Int, j: Int)Int
match expected type (?, ?) => ?
       Set(1,2) reduce X.f
                         ^
```

If I'm understanding things correctly, in this example the compiler is unable to use the fact that reduce expects a `(T, T) => T` **for `T >: Int`** to exclude the `f(s: String...)` option. Specifically, it seems that the type bound does not come into play, and the compiler is instead simply using `(T, T) => T` as the expected type?

**Comment by demobox** [comment]
See https://issues.scala-lang.org/browse/SI-9738?focusedCommentId=76231&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-76231

**Comment by demobox** [comment]
@som-snytt Just as I was getting to the point of feeling that I had a vague understanding of which examples might work, and which ones might not:

```
scala> Set(1L).reduce(math.max) // doesn't work, as we know
<console>:12: error: type mismatch;
 found   : (Int, Int) => Int
 required: (AnyVal, AnyVal) => AnyVal
       Set(1L).reduce(math.max)
                           ^

scala> def reduce[A, B >: A](s: Set[A])(fun: (B, B) => B): B = s.reduce(fun)
reduce: [A, B >: A](s: Set[A])(fun: (B, B) => B)B

scala> reduce(Set(1L))(math.max) // but then why does this work?!
res0: Long = 1
```

The problem the compiler would appear to have to solve looks tantalisingly similar: finding a version of math.max of type (B, B) => B, knowing that B >: Long. But this time, it can pick the right version:

```
        private[this] val res2: Long = $line27.$read.$iw.$iw.reduce[Long, Long](
scala.Predef.Set.apply[Long](1L))({
          ((x: Long, y: Long) => scala.math.`package`.max(x, y))
        });
```

So the problem the compiler is dealing with must be different somehow, or are the times in which this works vs. when it fails just generally not categorizable in any simple way?

**Comment by som-snytt** [comment]
Inference is left to right. I commented on the ticket:

-Ytyper-debug shows x.reduce(f) types f with type param to x.reduce unsolved; but reduce(x)(f) already determines the type args after the first arg list, so there is an expected type for f.

And Jira never annoyed me as much as a broken Jira.

**Comment by demobox** [comment]
> -Ytyper-debug shows x.reduce(f) types f with type param to x.reduce unsolved; but reduce(x)(f) already determines the type args after the first arg list, so there is an expected type for f.

If I'm looking at the appropriate part of the `typer-debug` output (hardly the expert, so this may be totally wrong), the difference is:

```
|    |    solving for (A: ?A)
|    |    \-> scala.collection.immutable.Set[Long]
|    solving for (A: ?A, B: ?B)
|    \-> (fun: (Long, Long) => Long)Long
```

vs. 

```
|    |    \-> scala.collection.immutable.Set[Long]
|    [adapt] [A1 >: A](op: (A1, A1) => A1)A1 adapted to
 [A1 >: A](op: (A1, A1) => A1)A1
|    \-> (op: (A1, A1) => A1)A1
...
solving for (A1: ?A1)
```

I.e. in the successful `reduce(x)(f)` case, it solves a type equation involving _both_ parameters, and succeeds with `B == Long`; in the unsuccessful case, it does not attempt to solve the `A1 >: A` type equation until _after_ overload resolution has completed, by which time the wrong method is chosen.

Thanks for taking yet another look at this, @som-snytt!

**Comment by som-snytt** [comment]
No, it's because `reduce(Set(1L))` already determines `B` because of the constraint, so the application `(f)` has an expected type `(Long, Long) => Long`. You can also see that `reduce(Set(1L))(_)` works for the same reason, and `vs.reduce(_)` does not.

If you remove the constraint, the function can only be `(Any, Any) => ???`; the function you supply can't help infer `B`.

They often suggest supplying a bogus type, to let the compiler error tell you what type is expected:
```
 reduce(Set(1L))((x: String, y: String) => (x+y).toLong)  // required: (Long, Long) => Long
 // without .toLong, required: Long
```

**Comment by demobox** [comment]
> No, it's because reduce(Set(1L)) already determines B because of the constraint [...] If you remove the constraint, the function can only be `(Any, Any) => ???`

Isn't the constraint present in both cases, though? At least, according the the method definition? My question perhaps wasn't phrased clearly: would it be accurate to say that, in the working case, the compiler solves the constraint system _before_ overload resolution, whereas in the failing case it only does to _after_ overload resolution?

And if that is correct, is that because, in the successful case, it is solving the constraint system as it needs to determine `A`, i.e. the type of the set?

PS: Thanks for sticking with this for so long - please feel free to drop it before we reach the point of exhaustion ;-)

**Comment by som-snytt** [comment]
I'll stick with it while my memory is refreshed; I'll forget all about it again soon.

Yes, you're not wrong. For some reason, I find it simpler just to say, after typing `reduce[A,B](vs)`, `B` is `Long` (because `A` is `Long` and `B >: A`); after typing `vs.reduce`, the type args haven't been solved yet, and can't guide the resolution of `max`.

(Sidebar: you could imagine taking each alternative of `max` in turn, inferring type args, and if only one such alternative typechecks, use it, or use the most specific among such alternatives. That's our intuition when we say, It's obvious which `max` I meant.)

This isn't too mysterious; it's why people like currying; it aids type inference.

**Comment by demobox** [comment]
> This isn't too mysterious; it's why people like currying; it aids type inference.

No, indeed, I think the mystery is not why the type inference helps here - the _apparent_ mystery is why the compiler "decides" to solve the type args in one case, but not in another that superficially seems similar. I.e.

```
def reduce[A, B >: A](s: Set[A])(fun: (B, B) => B): B = ???
```

doesn't to me look all that different from:

```
class R[A](val s: Set[A]) { def reduce[B >: A](fun: (B, B) => B): B = ??? }
```

Yet they are handled by the compiler in a sufficiently different fashion such that one succeeds, but the other does not.

**Comment by som-snytt** [comment]
I see what you meant before: in the first case, A and B are in the same type param section; in the second, failing case, A is a class type param, and B dangles on the method. Although in both cases you already know A, only in the first case does that also get you B.

**Comment by demobox** [comment]
> you already know A, only in the first case does that also get you B

Indeed - thanks for summarizing!

**Comment by demobox** [comment]
Closed via #149. See http://scalapuzzlers.com/#pzzlr-070.

Thanks, @som-snytt!
